lightning/src/ln/chanmon_update_fail_tests.rs

```diff
@@ -3089,3 +3098,127 @@ fn test_blocked_chan_preimage_release() {
 	do_commitment_signed_dance(&nodes[2], &nodes[1], &bs_htlc_fulfill_updates.commitment_signed, false, false);
 	expect_payment_sent(&nodes[2], payment_preimage_2, None, true, true);
 }
+
+fn do_test_inverted_mon_completion_order(complete_bc_commitment_dance: bool) {
+	// When we forward a payment and receive an `update_fulfill_htlc` message from the downstream
+	// channel, we immediately claim the HTLC on the upstream channel, before even doing a
+	// `commitment_signed` dance on the downstream channel. This implies that our
+	// `ChannelMonitorUpdate`s "go out" in the right order - first we ensure we'll get our money,
+	// then we write the update that resolves giving money on the downstream node. This is safe as
+	// long as `ChannelMonitorUpdate`s complete in the order in which they are generated, but of
+	// course this may not be the case. For asynchronous update writes, we have to ensure monitor
+	// updates can block each other, preventing the inversion all together.
+	let chanmon_cfgs = create_chanmon_cfgs(3);
+	let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
+
+	let persister;
+	let new_chain_monitor;
+	let nodes_1_deserialized;
+
+	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
+	let mut nodes = create_network(3, &node_cfgs, &node_chanmgrs);
+
+	let chan_id_ab = create_announced_chan_between_nodes(&nodes, 0, 1).2;
+	let chan_id_bc = create_announced_chan_between_nodes(&nodes, 1, 2).2;
+
+	// Route a payment from A, through B, to C, then claim it on C. Once we pass B the
+	// `update_fulfill_htlc` we have a monitor update for both of B's channels. We complete the one
+	// on the B<->C channel but leave the A<->B monitor update pending, then reload B.
```

Comment:
> Here we complete the B<->C one first, in the "wrong order"

Hmm... "complete" being "persist"?

I thought the point of the PR according to the description was to "enforce the correct ordering by blocking the downstream ChannelMonitorUpdate until the upstream one completes".

And a later comment says:

```rust
		// At this point node B still hasn't persisted the `ChannelMonitorUpdate` with the
		// preimage in the A <-> B channel, which will prevent it from persisting the
		// `ChannelMonitorUpdate` for the B<->C channel here to avoid "losing" the preimage.
```

Which seems to contradict "we complete the B<->C one first" if "complete" means "persist".

Sorry, just trying to get my head around this. From someone not as familiar with this part of the code, words like "complete" and "blocking" are vague. (Complete what? Block from doing what? Does claiming an HTLC involve persisting? etc.) Which makes it a little difficult to tie the comments to what's happening in the test.

Reply:
> Errr, okay, I see the confusion. Right, so its not the first ChannelMonitorUpdate that we block, its the second one.

First and second here are referring to the order in which the `ChannelMonitorUpdate`s are generated? Which do these correspond to?

Also, _what_ is blocked? Generating them? Persisting them? Updating the state machine?

> When we get an update_fulfill_htlc we generate monitor updates for the A<->B channel first (getting the preimage to disk on the channel where the preimage gets us money!). Then we generate the monitor update for the B<->C channel (also including the preimage). If the preimage update for the A<->B channel doesn't complete/make it to disk/persist, and we step the B<->C channel state machine forward a few times, we might get to a point where we may remove the preimage from the B<->C channel (case we don't need it for that channel anymore). At that time, the preimage for the A<->B channel better have made it to disk, otherwise we forget the preimage and lose the HTLC!
> 

The first sentence here says they are generated first for upstream (A<->B) and then for downstream (B<->C). 

> Thus, we don't block the first update on the B<->C channel, as that would delay something that's happening right now, but rather block the second one, which could remove the preimage we need.

But now you say the first update is on the downstream channel (B<->C) and the second one on the upstream channel (A<->B) is blocked.

And in the previous paragraph you say moving the B<->C channel forward could remove the preimage. But now you are saying that about the A<->B channel ("block the second one").

See my confusion? :)

Reply:
Ah, ok, thanks for explaining. My confusion was around two updates on B<->C, clearly, lol, and when each one was generated.

As a future project, it would be nice if the tests were more explicit about which monitor updates are expected on which channels rather than the more generic `check_added_monitors`.